File: 079WaterArea/waterArea.py
# Solution 1
# O(n) time / O(n) space

def waterArea(heights):
    maxes = [0 for x in heights]

    leftMax = 0
    for i in range(len(heights)):
        height = heights[i]
        maxes[i] = leftMax
        leftMax = max(leftMax, height)

    rightMax = 0
    for i in reversed(range(len(heights))):
        height = heights[i]
        minHeight = min(rightMax, maxes[i])
        if height < minHeight:
            maxes[i] = minHeight - height
        else:
            maxes[i] = 0
        rightMax = max(rightMax, height)
    return sum(maxes)

# Solution 2
# O() time / O() space


def waterArea(heights):
    if len(heights) == 0:
        return 0

    leftIdx = 0
    rightIdx = len(heights) - 1
    leftMax = heights[leftIdx]
    rightMax = heights[rightIdx]
    waterSurface = 0

    while leftIdx < rightIdx:
        if heights[leftIdx] < heights[rightIdx]:
            leftIdx += 1
            leftMax = max(leftMax, heights[leftIdx])
            waterSurface += leftMax - heights[leftIdx]
        else:
            rightIdx -= 1
            rightMax = max(rightMax, heights[rightIdx])
            waterSurface += rightMax - heights[rightIdx]

    return waterSurface

File: 079WaterArea/water_area.rb
# Solution 1
# O(n) time / O(n) space

def water_area(heights)
  maxes = Array.new(heights.length, 0)

  left_max = 0
  (0...heights.length - 1).each do |i|
    height = heights[i]
    maxes[i] = left_max
    left_max = [left_max, height].max
  end

  right_max = 0
  (heights.length - 1...0).each do |i|
    height = heights[i]
    min_height = [right_max, maxes[i]].min
    maxes[i] = if height < min_height
                 min_height - height
               else
                 0
               end
    right_max = [right_max, height].max
  end
  p maxes
  maxes.sum
end

# Solution 2
# O(n) time / O(1) space
